### ISSUE TITLE
Metadata is not saved with object using Python SDK

### ISSUE THREAD

**Issue by alirezastack** [minio/minio#6027]
## Expected Behavior
* Metadata should be saved with object.

## Current Behavior
* Given metadata is not stored:

```
metadata = {
                'user_id': user_id,
                'original_name': image_file.filename,
                'file_format': file_format
}
out = self.minio_client.fput_object(bucket_name=buck_name,
                                                object_name=file_name,
                                                file_path=image_thumbnail_path,
                                                metadata=metadata)
```

Object is stored successfully, now when I get metadata information as below:
```
# Get a full object
data = self.minio_client.get_object(buck_name, file_name)
# Get metadata information
obj = self.minio_client.stat_object(buck_name, file_name)
app.logger.debug(vars(obj))
```

Last line that logs information reports that:

`{'last_modified': time.struct_time(tm_year=2018, tm_mon=6, tm_mday=7, tm_hour=20, tm_min=14, tm_sec=17, tm_wday=3, tm_yday=158, tm_isdst=0), 'object_name': u'e154bd767abf463d9cc5a26319c9c805.jpg', 'bucket_name': 'buck-name', 'etag': '1e981f43ef54caaebd87ccf39ac6b17f', 'content_type': 'application/octet-stream', 'is_dir': False, 'metadata': {}, 'size': 40673}`

## Your Environment
* Version used (minio version): DOCKER image with latest version
* Minio Python SDK: 4.0.0
* VERSION: 2018-05-25T19:49:13Z
* MEMORY: Used: 5.2 MB | Allocated: 20 MB | Used-Heap: 5.2 MB | Allocated-Heap: 8.7 MB
* PLATFORM Host: cf124cb3e9bb | OS: linux | Arch: amd64
* RUNTIME: Version: go1.10.1 | CPUs: 2

**Comment by deekoder** [comment]
Will look into it @alirezastack

**Comment by araujof** [comment]
I think I'm observing a related, but different problem. In my case, I pushed some objects with additional metadata into minio, which I can verify using the mc client:
```
# mc stat minio/monitoringdata/mon.1528743864
Name      : mon.1528743864
Date      : 2018-06-11 19:04:38 UTC 
Size      : 181KiB 
ETag      : e58e950b0e95aaa9c2b89087923865e9 
Type      : file 
Metadata  :
  X-Minio-Meta-Trace: complete 
  Content-Type      : application/octet-stream
``` 
However, when trying to stat the same object via the Python SDK, the metadata dictionary returns empty. Here is a printout of the `obj` returned by the Python client:
```
2018-06-11 15:05:05,287 [INFO]    <Object: bucket_name: monitoringdata object_name: mon.1528743864 last_modified: time.struct_time(tm_year=2018, tm_mon=6, tm_mday=11, tm_hour=19, tm_min=4, tm_sec=38, tm_wday=0, tm_yday=162, tm_isdst=0) etag: e58e950b0e95aaa9c2b89087923865e9 size: 185629 content_type: application/octet-stream, is_dir: False, metadata: {}>
```
Is this a bug in the Python SDK, or I'm missing something? Any thoughts?

**Comment by alirezastack** [comment]
Any update on the issue? Is the bug confirmed? Should I ask this on `minio-py` repo?

**Comment by araujof** [comment]
I glanced over the Python SDK code, and it looks like this is due to undocumented behavior.

When stating an object, the `stat_object` API performs the following check on the returned headers:
```
## Capture only custom metadata.
custom_metadata = dict()
for k in response.headers:
    if is_supported_header(k) or is_amz_header(k):
        custom_metadata[k] = response.headers.get(k)
```
Looking at the helper function definitions, I now understand why my `X-Minio-Meta-Trace` metadata wasn't returned in the dictionary. However, this object (and metadata) was saved with the `fput_object` API, and accepted my metadata definition, as I can tell by querying minio with `mc`. In addition, this behavior is inconsistent with `mc` itself. 
```
# returns true if amz s3 system defined metadata
def is_amz_header(key):
    key = key.lower()
    return key.startswith("x-amz-meta") or key == "x-amz-acl" or key.startswith("x-amz-server-side-encryption")

# returns true if a standard supported header
def is_supported_header(key):
    ## Supported headers for object.
    supported_headers = [
	   "cache-control",
	   "content-encoding",
	   "content-disposition",
	   "content-language",
	   "x-amz-website-redirect-location",
            ## Add more supported headers here.
        ]
    return key.lower() in supported_headers
```
Shouldn't this check be loosened a little bit and accept custom metadata definitions (perhaps in the form of `X-Minio-<something>`. Of course, this has to be documented in the SDK docs :-)

The workaround, for now, seems to define metadata using `X-Amz` prefix.

**Comment by araujof** [comment]
https://github.com/minio/minio-py/issues/662

**Comment by ebozduman** [comment]
I reproduced the issue with FS minio server. I am looking into it.

**Comment by ebozduman** [comment]
@alirezastack, @araujof,
Missing custom meta data issue has been already fixed in the latest [Python SDK 4.0.1 release](https://github.com/minio/minio-py/releases). 
```
Changelog
:
:
api:put_object not saving user metadata without x-amz-meta- prefix (#655) (#656) (05/02/18)
```
So, please upgrade and try it out again.

FYI:
I still see a similar problem with `"x-amz-storage-class"`metadata and that's why I thought I reproduced the issue with the latest  Python SDK, 4.0.1. I'll be still looking into it.
Let us know if you still have questions.

**Comment by ebozduman** [comment]
FYI
About `"x-amz-storage-class"` metadata tag: aws s3 also behaves the same way. Stat command does not return/support this metadata tag. It is only meaningful when putting objects. So, Minio is compatible with aws s3.

**Comment by araujof** [comment]
@ebozduman,

Thanks. As I explained in my previous comment, the custom metadata is stored if I define it with prefix `x-amz-meta`, _or_ even with prefix `x-mino-meta`, through the Python SDK function `fput_object`; the issue we describe arises when reading (stating) the metadata via the SDK: due to the check `is_amz_header(key)`, the custom metadata is not added to the metadata dictionary returned to the user. 

I understand that this (x-minio-meta) may not be a metadata prefix supported by S3, but the Python SDK should document this (or at least raise an error/log a warning message when an attempt to store the unsupported metadata occurs). Maybe I'm missing something, but from what I can tell, Minio could fully support S3 while allowing other metadata schemes to be stored in the object store. If strict compliance to S3 is what is desired, then the documentation could be improved to highlight the enforced input validation (and perhaps the `fput_object` and `put_object` methods could log/alert the user about the contract violation).

**Comment by ebozduman** [comment]
@araujof,
But our fix in Python SDK 4.0.1 already appends the prefix `x-amz-meta` if custom/unsupported metadata is provided by the caller. Please see the  change log excerpt I added in the above comment, or the latest version of the helpers.py file.
```
# return a new metadata dictionary where user defined metadata keys
# are prefixed by "x-amz-meta-"
def amzprefix_user_metadata(metadata):
    m = dict()
    for k,v in metadata.items():
       if is_amz_header(k) or is_supported_header(k) or is_storageclass_header(k):
            print("amzprefix_user_metadata>> Adding it to m", k, v)
            m[k] = v
       else:
       m["X-Amz-Meta-" + k] = v         #  <<<<<<<<<< prepending <<<<<<<<<<<<<
    return m
```

So, if metadata is provided as:
```
metadata = {
    'user_id': "user_id",
    'original_name': "image_file.filename",
    'file_format': "file_format"
}
```
the output of stat_object api returns the newly renamed custom metadata:
```
>>> bucket_name="mybucket"
>>> object_name="testfile"
>>> client.fput_object(bucket_name, object_name+'-f', testfile, metadata=metadata)
'e2835de400ca762435ae29de87dd55be'
>>> resp = client.stat_object(bucket_name, object_name+'-f')
>>> print(resp)
<Object: bucket_name: mybucket object_name: b'testfile-f' last_modified: time.struct_time(tm_year=2018, tm_mon=6, tm_mday=14, tm_hour=17, tm_min=48, tm_sec=8, tm_wday=3, tm_yday=165, tm_isdst=0) 
etag: bbc0b92a7c4ad57451fb7b6a479b6735 
size: 1048576 
content_type: application/octet-stream, 
is_dir: False, 
metadata: {
    'X-Amz-Meta-File_format': 'file_format',
    'X-Amz-Meta-Original_name': 'image_file.filename', 
    'X-Amz-Meta-User_id': 'user_id'}>
>>> 
```
We definitely need to update our Python SDK docs for the added prefix, `X-Amz-Meta-`, when custom metadata is provided by the user.
Other than that, I believe current behavior satisfies the concerns/problems you were talking about.
Please confirm or clarify if I am missing something.

**Comment by harshavardhana** [comment]
> Other than that, I believe current behavior satisfies the concerns/problems you were talking about.
> Please confirm or clarify if I am missing something.

Closing this issue as fixed for now.
